### PR TITLE
Support basic Opera docker node

### PR DESCRIPTION
**Pull request by Drimix20**
I would like to use opera node in selenium docker hub. I think that it can be useful also for others, so I created PR to be able to make these changes official and not only for private usage.

I don't have any experience with automation which is created in SeleniumHQ so I tried to make changes in all required other files too, not only created NodeOpera directory.

I tested creation of opera selenium image and it was correctly connected to running selenium docker hub.
I'll provide screenshots and some details about scripts.

- [X] By placing an `X` in the preceding checkbox, I verify that I have signed the [Contributor License Agreement](https://github.com/SeleniumHQ/docker-selenium/blob/master/CONTRIBUTING.md#contributing-code-to-selenium)

### CODE REVIEW

**Comment by manuferna1985** [comment]
I have been configuring a selenium hub using Chrome and Firefox nodes recently, and all this official docker images are simply great! Good work guys! But like Drimix20, I think that if we (me the first) contribute with other images for Opera, Safari, and the other most used browsers that are available for Unix, the % of browsers usage that we could cover would be nearly 90% (ex Edge/IE...).

There is any technical impediment right now to support this project with a Safari nose for example?

**Comment by Drimix20** [comment]
> 
> 
> @Drimix20 Could you please fix the merge conflicts first? After that we can review the PR.

I performed rebase.

**Comment by Drimix20** [comment]
> @manuferna1985
> 
> I have been configuring a selenium hub using Chrome and Firefox nodes recently, and all this official docker images are simply great! Good work guys! But like Drimix20, I think that if we (me the first) contribute with other images for Opera, Safari, and the other most used browsers that are available for Unix, the % of browsers usage that we could cover would be nearly 90% (ex Edge/IE...).
 
> There is any technical impediment right now to support this project with a Safari node for example?
As far as I know there is limitation based on host os system. Currently I don't know how to create docker image with OSX. I saw some hints like "create docker container, install emulator and emulate OSX". But I think this is not good way how to create docker image. Docker image should be lightweight.

File: Makefile [hunk 1]
```diff
@@ -108,6 +116,8 @@ release_latest:
 	docker push $(NAME)/node-base:latest
 	docker push $(NAME)/node-chrome:latest
 	docker push $(NAME)/node-firefox:latest
+	docker push $(NAME)/node-opera:latest
+	docker push $(NAME)/node-phantomjs:latest
```
Comment (diemol) [inline Makefile]: Why is `node-phantomjs` here? This image was deprecated a while ago.

File: Makefile [hunk 2]
```diff
@@ -121,6 +131,8 @@ tag_major_minor:
 	docker tag $(NAME)/node-base:$(VERSION) $(NAME)/node-base:$(MAJOR)
 	docker tag $(NAME)/node-chrome:$(VERSION) $(NAME)/node-chrome:$(MAJOR)
 	docker tag $(NAME)/node-firefox:$(VERSION) $(NAME)/node-firefox:$(MAJOR)
+	docker tag $(NAME)/node-opera:$(VERSION) $(NAME)/node-opera:$(MAJOR)
+	docker tag $(NAME)/node-phantomjs:$(VERSION) $(NAME)/node-phantomjs:$(MAJOR)
```
Comment (diemol) [inline Makefile]: Why is `node-phantomjs` here? This image was deprecated a while ago.

File: Makefile [hunk 3]
```diff
@@ -132,6 +144,8 @@ tag_major_minor:
 	docker tag $(NAME)/node-base:$(VERSION) $(NAME)/node-base:$(MAJOR).$(MINOR)
 	docker tag $(NAME)/node-chrome:$(VERSION) $(NAME)/node-chrome:$(MAJOR).$(MINOR)
 	docker tag $(NAME)/node-firefox:$(VERSION) $(NAME)/node-firefox:$(MAJOR).$(MINOR)
+	docker tag $(NAME)/node-opera:$(VERSION) $(NAME)/node-opera:$(MAJOR).$(MINOR)
+	docker tag $(NAME)/node-phantomjs:$(VERSION) $(NAME)/node-phantomjs:$(MAJOR).$(MINOR)
```
Comment (diemol) [inline Makefile]: Why is `node-phantomjs` here? This image was deprecated a while ago.

File: Makefile [hunk 4]
```diff
@@ -143,6 +157,8 @@ tag_major_minor:
 	docker tag $(NAME)/node-base:$(VERSION) $(NAME)/node-base:$(MAJOR_MINOR_PATCH)
 	docker tag $(NAME)/node-chrome:$(VERSION) $(NAME)/node-chrome:$(MAJOR_MINOR_PATCH)
 	docker tag $(NAME)/node-firefox:$(VERSION) $(NAME)/node-firefox:$(MAJOR_MINOR_PATCH)
+	docker tag $(NAME)/node-opera:$(VERSION) $(NAME)/node-opera:$(MAJOR_MINOR_PATCH)
+	docker tag $(NAME)/node-phantomjs:$(VERSION) $(NAME)/node-phantomjs:$(MAJOR_MINOR_PATCH)
```
Comment (diemol) [inline Makefile]: Why is `node-phantomjs` here? This image was deprecated a while ago.

File: Makefile [hunk 5]
```diff
@@ -156,6 +172,8 @@ release: tag_major_minor
 	@if ! docker images $(NAME)/node-base | awk '{ print $$2 }' | grep -q -F $(VERSION); then echo "$(NAME)/node-base version $(VERSION) is not yet built. Please run 'make build'"; false; fi
 	@if ! docker images $(NAME)/node-chrome | awk '{ print $$2 }' | grep -q -F $(VERSION); then echo "$(NAME)/node-chrome version $(VERSION) is not yet built. Please run 'make build'"; false; fi
 	@if ! docker images $(NAME)/node-firefox | awk '{ print $$2 }' | grep -q -F $(VERSION); then echo "$(NAME)/node-firefox version $(VERSION) is not yet built. Please run 'make build'"; false; fi
+	@if ! docker images $(NAME)/node-opera | awk '{ print $$2 }' | grep -q -F $(VERSION); then echo "$(NAME)/node-opera version $(VERSION) is not yet built. Please run 'make build'"; false; fi
+	@if ! docker images $(NAME)/node-phantomjs | awk '{ print $$2 }' | grep -q -F $(VERSION); then echo "$(NAME)/node-phantomjs version $(VERSION) is not yet built. Please run 'make build'"; false; fi
```
Comment (diemol) [inline Makefile]: Why is `node-phantomjs` here? This image was deprecated a while ago.

File: Makefile [hunk 6]
```diff
@@ -167,6 +185,8 @@ release: tag_major_minor
 	docker push $(NAME)/node-base:$(VERSION)
 	docker push $(NAME)/node-chrome:$(VERSION)
 	docker push $(NAME)/node-firefox:$(VERSION)
+	docker push $(NAME)/node-opera:$(VERSION)
+	docker push $(NAME)/node-phantomjs:$(VERSION)
```
Comment (diemol) [inline Makefile]: Why is `node-phantomjs` here? This image was deprecated a while ago.

File: Makefile [hunk 7]
```diff
@@ -178,6 +198,8 @@ release: tag_major_minor
 	docker push $(NAME)/node-base:$(MAJOR)
 	docker push $(NAME)/node-chrome:$(MAJOR)
 	docker push $(NAME)/node-firefox:$(MAJOR)
+	docker push $(NAME)/node-opera:$(MAJOR)
+	docker push $(NAME)/node-phantomjs:$(MAJOR)
```
Comment (diemol) [inline Makefile]: Why is `node-phantomjs` here? This image was deprecated a while ago.

File: Makefile [hunk 8]
```diff
@@ -189,6 +211,8 @@ release: tag_major_minor
 	docker push $(NAME)/node-base:$(MAJOR).$(MINOR)
 	docker push $(NAME)/node-chrome:$(MAJOR).$(MINOR)
 	docker push $(NAME)/node-firefox:$(MAJOR).$(MINOR)
+	docker push $(NAME)/node-opera:$(MAJOR).$(MINOR)
+	docker push $(NAME)/node-phantomjs:$(MAJOR).$(MINOR)
```
Comment (diemol) [inline Makefile]: Why is `node-phantomjs` here? This image was deprecated a while ago.

File: Makefile [hunk 9]
```diff
@@ -200,6 +224,8 @@ release: tag_major_minor
 	docker push $(NAME)/node-base:$(MAJOR_MINOR_PATCH)
 	docker push $(NAME)/node-chrome:$(MAJOR_MINOR_PATCH)
 	docker push $(NAME)/node-firefox:$(MAJOR_MINOR_PATCH)
+	docker push $(NAME)/node-opera:$(MAJOR_MINOR_PATCH)
+	docker push $(NAME)/node-phantomjs:$(MAJOR_MINOR_PATCH)
```
Comment (diemol) [inline Makefile]: Why is `node-phantomjs` here? This image was deprecated a while ago.

File: Makefile [hunk 10]
```diff
@@ -251,11 +281,15 @@ test_firefox_standalone_debug:
 	ci \
 	firefox \
 	firefox_debug \
+	opera \
+	phantomjs \
```
Comment (diemol) [inline Makefile]: Why is `node-phantomjs` here? This image was deprecated a while ago.

File: Makefile [hunk 11]
```diff
@@ -251,11 +281,15 @@ test_firefox_standalone_debug:
...
 	generate_all \
 	generate_hub \
 	generate_nodebase \
 	generate_chrome \
 	generate_firefox \
+	generate_opera \
+	generate_phantomjs \
```
Comment (diemol) [inline Makefile]: Why is `node-phantomjs` here? This image was deprecated a while ago.

File: NodeOpera/Dockerfile.txt [hunk 12]
```diff
@@ -0,0 +1,42 @@
+USER root
+
+#=========
+# Opera
+#=========
+ENV OPERA_VERSION=60.0.3255.170
```
Comment (diemol) [inline NodeOpera/Dockerfile.txt]: Why do we need to hardcode the Opera version, can't the latest one be retrieved dynamically? Can we install any given version?

File: NodeOpera/Dockerfile.txt [hunk 13]
```diff
@@ -0,0 +1,42 @@
...
+ENV OPERA_DRIVER_VERSION=v.2.45
```
Comment (diemol) [inline NodeOpera/Dockerfile.txt]: Do we need to hardcode the driver version?

File: NodeOpera/generate_config [hunk 14]
```diff
@@ -0,0 +1,26 @@
+#!/bin/bash
+
+OPERA_VERSION=60.0.3255.170
```
Comment (diemol) [inline NodeOpera/generate_config]: Why do we need to hardcode the Opera version?

File: Test/smoke-opera.js [hunk 15]
```diff
@@ -0,0 +1 @@
+require('./smoke-test')('opera');
```
Comment (diemol) [inline Test/smoke-opera.js]: There are the old test files when JavaScript was used, please let's use only Python.

File: sa-test.sh [hunk 16]
```diff
@@ -0,0 +1,42 @@
+#!/usr/bin/env bash
```
Comment (diemol) [inline sa-test.sh]: There are the old test files when JavaScript was used, please let's use only Python.

File: test.sh [hunk 17]
```diff
@@ -0,0 +1,103 @@
+#!/usr/bin/env bash
```
Comment (diemol) [inline test.sh]: There are the old test files when JavaScript was used, please let's use only Python.

File: tests/test.py [hunk 18]
```diff
@@ -36,6 +36,9 @@
     'NodeFirefoxDebug': 'node-firefox-debug',
     'StandaloneFirefox': 'standalone-firefox',
     'StandaloneFirefoxDebug': 'standalone-firefox-debug',
+
```
Comment (diemol) [inline tests/test.py]: Can you please add tests for the Standalone mode as well?

File: NodeOpera/Dockerfile.txt [hunk 19]
```diff
@@ -0,0 +1,42 @@
...
+
+RUN wget -O- http://deb.opera.com/archive.key | apt-key add - && \
+     echo 'deb http://deb.opera.com/opera-stable/ stable non-free' > /etc/apt/sources.list.d/opera-blink.list
+
+RUN apt-get update
+
+RUN apt-get update && \
```
Comment (cgoldberg) [inline NodeOpera/Dockerfile.txt]: don't need to run `apt-get update` again.

File: NodeOpera/Dockerfile.txt [hunk 20]
```diff
@@ -0,0 +1,42 @@
...
+     apt-get -y --no-install-recommends install opera-stable && \
+     opera --version && \
+     rm -Rf /tmp/* && \
```
Comment (cgoldberg) [inline NodeOpera/Dockerfile.txt]: why do you need to clean /tmp?

**Comment by Drimix20** [comment]
@diemol So should I also create standalone and debug version? I created only node version without debug. Thanks for review, I'll resolve all comments.

**Comment by diemol** [comment]
@Drimix20 yes, ideally Opera should have images for debug and standalone as well.

File: Makefile (hunk 1 above)
Comment (Drimix20) [reply Makefile]: Deleted.

File: Makefile (hunk 2 above)
Comment (Drimix20) [reply Makefile]: Deleted

File: Makefile (hunk 3 above)
Comment (Drimix20) [reply Makefile]: Deleted

File: Makefile (hunk 4 above)
Comment (Drimix20) [reply Makefile]: Deleted

File: test.sh (hunk 17 above)
Comment (Drimix20) [reply test.sh]: Ok, deleted.

File: sa-test.sh (hunk 16 above)
Comment (Drimix20) [reply sa-test.sh]: Ok, deleted.

File: Test/smoke-opera.js (hunk 15 above)
Comment (Drimix20) [reply Test/smoke-opera.js]: Ok, deleted.

File: NodeOpera/generate_config (hunk 14 above)
Comment (Drimix20) [reply NodeOpera/generate_config]: This hardcoded version is supported by used opera chromium driver. Should I use different approach?

File: NodeOpera/Dockerfile.txt (hunk 19 above)
Comment (Drimix20) [reply NodeOpera/Dockerfile.txt]: Deleted.

File: NodeOpera/Dockerfile.txt (hunk 20 above)
Comment (Drimix20) [reply NodeOpera/Dockerfile.txt]: Deleted. It's leftover from different approach how to install opera.

**Comment by Drimix20** [comment]
@diemol Yes, I'm already working on others types of docker images. I'm also resolving all comments. Thanks for feedback!

File: NodeOpera/generate_config (hunk 14 above)
Comment (diemol) [reply NodeOpera/generate_config]: Please check if the same approach used for Chrome works here, which is installing the latest available package from the distribution. 
The idea is that a browser upgrade should be just running a new build.

File: Makefile (hunk 5 above)
Comment (Drimix20) [reply Makefile]: Deleted.

File: Makefile (hunk 6 above)
Comment (Drimix20) [reply Makefile]: Deleted.

File: Makefile (hunk 7 above)
Comment (Drimix20) [reply Makefile]: Deleted.

File: Makefile (hunk 8 above)
Comment (Drimix20) [reply Makefile]: Deleted.

File: Makefile (hunk 9 above)
Comment (Drimix20) [reply Makefile]: Deleted.

File: Makefile (hunk 10 above)
Comment (Drimix20) [reply Makefile]: Deleted.

File: Makefile (hunk 11 above)
Comment (Drimix20) [reply Makefile]: Deleted.

File: NodeOpera/Dockerfile.txt (hunk 12 above)
Comment (Drimix20) [reply NodeOpera/Dockerfile.txt]: Env variable opera version deleted. Always latest available opera will be installed.

File: NodeOpera/Dockerfile.txt (hunk 13 above)
Comment (Drimix20) [reply NodeOpera/Dockerfile.txt]: Deleted. I will use link to latest release instead, so hardcoded version is not required.

File: tests/test.py (hunk 18 above)
Comment (Drimix20) [reply tests/test.py]: Added.

File: NodeOpera/generate_config (hunk 14 above)
Comment (Drimix20) [reply NodeOpera/generate_config]: Ok, I removed all hardcoded versions and I made it dynamic based on approach used for Chrome docker images.

**Comment by Drimix20** [comment]
I added all opera images, I performed rebase and updated tests for other type of images. But right know I'm waiting for PR build, so when it will finish, I'll continue with solving problems with failing tests.

**Comment by ddavison** [comment]
related: https://github.com/SeleniumHQ/docker-selenium/pull/791#issuecomment-428666838

I believe we decided against Opera as it uses Chromium as its engine. `NodeChrome` already uses Chromium

**Comment by Drimix20** [comment]
@ddavison  Sorry, I missed your PR, but I was looking for opera keyword.
@diemol Can you please reply?

**Comment by diemol** [comment]
@Drimix20 I assumed that you double checked for previous PRs/issues 😄 

I've seen this request for Opera a few times, so instead of saying no this time, I think we should take the approach of accepting the PR, adding Opera, measuring the usage. 
If the images are actually used and helpful, we could continue to support them, but if after a few months we don't see any high usage, we could just remove them.

So far, this PR is not in a good shape, so I would leave it to your judgement @Drimix20 if it makes sense to continue based on the fact that Opera is based on Chromium.

**Comment by Drimix20** [comment]
Ok, I will finish the PR based on approach "accepting the PR, adding Opera, measuring the usage". 

PR is not in good shape because I'm facing to problem with opening new opera session. I found out that there is reported issue https://github.com/operasoftware/operachromiumdriver/issues/69 on operachromiumdriver, that opera version 62.* is not supported by latest operachromiumdriver (2.25). I'm trying to install older 60.* opera version and then I'll see if there is another issue.

I created all type of images so this should be the last missing and important thing.

**Comment by diemol** [comment]
Thanks for referecing the issue @Drimix20, which kind of worries me because I would not like to add a browser image if the dependencies are not properly maintained. In this case we cannot use Opera 62 because the driver is not up to date. Maybe this is an exception and the driver is usually well maintained, but this looks like a blocker.

**Comment by Drimix20** [comment]
Hi, operachromedriver is released and is working with Opera 62 version. I updated dockerfile, but I'm still getting error "No matching capabilities found". Can you please navigate me, what can be wrong? 

I've created selenium hub with chrome and opera node. Both nodes were correctly connected to selenium hub. I see also configured capabilities, but I'm not able to find out, why it is not working properly.

Selenium hub with nodes:
image


Trying to run script:
`
from selenium import webdriver
from selenium.webdriver.common.desired_capabilities import DesiredCapabilities

driver = webdriver.Remote(
        command_executor='http://127.0.0.1:4444/wd/hub',
        desired_capabilities=DesiredCapabilities.OPERA)
driver.get('http://the-internet.herokuapp.com/nested_frames')
`

**Pull request closed without merging**